azure_sdk_storage_core/src/connection_string.rs

```diff
@@ -0,0 +1,348 @@
+use azure_sdk_core::errors::AzureError;
+
+// Key names.
+pub const ACCOUNT_KEY_KEY_NAME: &str = "AccountKey";
+pub const ACCOUNT_NAME_KEY_NAME: &str = "AccountName";
+pub const SAS_KEY_NAME: &str = "SharedAccessSignature";
+pub const ENDPOINT_SUFFIX_KEY_NAME: &str = "EndpointSuffix";
+pub const DEFAULT_ENDPOINTS_PROTOCOL_KEY_NAME: &str = "DefaultEndpointsProtocol";
+pub const USE_DEVELOPMENT_STORAGE_KEY_NAME: &str = "UseDevelopmentStorage";
+pub const DEVELOPMENT_STORAGE_PROXY_URI_KEY_NAME: &str = "DevelopmentStorageProxyUri";
+pub const BLOB_ENDPOINT_KEY_NAME: &str = "BlobEndpoint";
+pub const BLOB_SECONDARY_ENDPOINT_KEY_NAME: &str = "BlobSecondaryEndpoint";
+pub const TABLE_ENDPOINT_KEY_NAME: &str = "TableEndpoint";
+pub const TABLE_SECONDARY_ENDPOINT_KEY_NAME: &str = "TableSecondaryEndpoint";
+pub const QUEUE_ENDPOINT_KEY_NAME: &str = "QueueEndpoint";
+pub const QUEUE_SECONDARY_ENDPOINT_KEY_NAME: &str = "QueueSecondaryEndpoint";
+pub const FILE_ENDPOINT_KEY_NAME: &str = "FileEndpoint";
+pub const FILE_SECONDARY_ENDPOINT_KEY_NAME: &str = "FileSecondaryEndpoint";
+
+quick_error! {
+    #[derive(Debug)]
+    pub enum ConnectionStringError {
+        MissingValue { key: String } {
+            display("Missing value for key '{}'", key)
+        }
+        UnexpectedKey { key: String } {
+            display("Unexpected for key '{}'", key)
+        }
+        ParsingError { msg: String } {
+            display("{}", msg)
+        }
+    }
+}
+
+impl From<ConnectionStringError> for AzureError {
+    fn from(err: ConnectionStringError) -> Self {
+        AzureError::GenericErrorWithText(err.to_string())
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum EndpointProtocol {
+    Http,
+    Https,
+}
+
+impl std::fmt::Display for EndpointProtocol {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        let s = match self {
+            EndpointProtocol::Https => "https",
+            EndpointProtocol::Http => "http",
+        };
+        write!(f, "{}", s)
+    }
+}
+
+/// A storage connection string.
+///
+/// The key are a subset of what is defined in the
+/// [.NET SDK](https://github.com/Azure/azure-storage-net/blob/ed89733dfb170707d65b7b8b0be36cb5bd6512e4/Lib/Common/CloudStorageAccount.cs#L63-L261).
+#[derive(Debug)]
+pub struct ConnectionString<'a> {
```

Comment:
Missing some keys here, but I think these are the more popular ones.
Let me know if this is a blocker @MindFlavor.

As a side note, I considered doing a loose mapping here (basically making this a wrapper around a dictionary) and for example making `cs.account_name()` a method that returns the `AccountName` key from the mapping.
The pro is lazy-fetching of the properties, explicitness, and easier maintenance of these properties; the cons are an expensive `HashMap` and only getting errors upon key retrieval.

Could go either way here, not critical I think.

Reply:
No it is not! It's great as it is now. We can always come back and add them in the future. As for the `HashMap` I prefer your approach: finding out an invalid connection string as soon as possible is important. Also, your code is elegant :+1: !